km/km_coredump.c

```diff
@@ -626,6 +639,157 @@ static inline int km_guestmem_write(int fd, km_gva_t base, size_t length)
    return 0;
 }
 
+/*
+ * The following macros define the format of the 8 byte words in
+ * the file /proc/PID/pagemap.
+ * They were copied from linux kernel source file fs/proc/task_mmu.c
+ * So far only km_guestmem_write_sparse() needs these.
+ */
+#define BIT_ULL(nr) (1ULL << (nr))
+#define GENMASK_ULL(h, l) (((~0ULL) << (l)) & (~0ULL >> (sizeof(long long) * 8 - 1 - (h))))
+#define PM_PFRAME_BITS 55
+#define PM_PFRAME_MASK GENMASK_ULL(PM_PFRAME_BITS - 1, 0)
+#define PM_SOFT_DIRTY BIT_ULL(55)
+#define PM_MMAP_EXCLUSIVE BIT_ULL(56)
+#define PM_UFFD_WP BIT_ULL(57)
+#define PM_FILE BIT_ULL(61)
+#define PM_SWAP BIT_ULL(62)
+#define PM_PRESENT BIT_ULL(63)
+
+/*
+ * Only write virtual memory pages that have a physical memory page or
+ * a page on backing store.
+ * Any other pages in the virtual address space have no backing and will
+ * be created with zero fill when they are referenced.
+ * Since the pages have nothing of value in them we don't need to save them
+ * in our snapshots.
+ * We leave holes in the produced core file for these unbacked pages since
+ * the holes will read as zero filled bytes.
+ * A little investigation seemed to show that the memory manager in the kernel
+ * is willing to create unbacked pages when chunks of the core file are mapped
+ * into memory with mmap() so, we don't need to do anything special during snapshot
+ * startup to ensure the holes in the snapshot file appear as unbacked memory
+ * pages in the start payload.
+ * It should also be noted that the linux cp and tar programs are willing to detect
+ * holes in files and ensure the holes are propagated when copies are made.
+ * Use their --sparse command line flags.
+ */
+#define PROC_PAGEMAP "/proc/self/pagemap"
+#define PROC_PAGEMAP_ENTRYSIZE (sizeof(uint64_t))
+static inline int km_guestmem_write_sparse(int corefd, km_gva_t base, size_t length)
+{
+   km_assert(base == roundup(base, KM_PAGE_SIZE));
+
+   // align to page boundary in output file
+   off_t base_offset = 0;
+   int rc = km_corefile_align(corefd, &base_offset);
+   if (rc != 0) {
+      return rc;
+   }
+
+   km_infox(KM_TRACE_COREDUMP, "base=0x%lx length=0x%lx file offset=0x%lx", base, length, base_offset);
+
+   // !!!! use km virtual address
+   km_kma_t kma_base = km_gva_to_kma(base);
+
+   // Get virtual memory page backing info for the passed memory range from /proc/self/pagemap
+   size_t bytestowrite = roundup(length, KM_PAGE_SIZE);
+   size_t numberofpages = bytestowrite / KM_PAGE_SIZE;
+   uint64_t* pagemap = malloc(numberofpages * PROC_PAGEMAP_ENTRYSIZE);
+   int pagemapfd = open(PROC_PAGEMAP, O_RDONLY);
```

Comment:
This function gets called several times. open/close can be moved out. But that is just an opinion.